R/preprocess.R

```diff
@@ -199,63 +244,101 @@ enw_extend_date <- function(metaobs, days = 20, direction = "end") {
   exts[, observed := FALSE]
 
   exts <- rbind(
-    data.table::copy(metaobs)[, observed := TRUE],
+    metaobs[, observed := TRUE],
     exts[, observed := FALSE]
   )
   data.table::setkeyv(exts, c(".group", "date"))
   return(exts[])
 }
 
-#' @title FUNCTION_TITLE
-#' @description FUNCTION_DESCRIPTION
-#' @param obs PARAM_DESCRIPTION
-#' @param by PARAM_DESCRIPTION, Default: c()
-#' @return OUTPUT_DESCRIPTION
+#' @title Assign a group to each row of a data.table
+#' 
+#' @description Assign a group to each row of a data.table. If `by` is
+#' specified, then each unique combination of the columns in `by` will
+#' be assigned a unique group. If `by` is not specified, then all rows
+#' will be assigned to the same group.
+#'
+#' @param obs A data.table or data.frame without a `.group` column.
+#'
+#' @param by A character vector of column names to group by. Defaults to
+#' an empty vector.
+#'
+#' @return A data.table with a `.group` column added ordered by `.group`
+#' and the existing key of `obs`.
+#'
 #' @family preprocess
 #' @export
 #' @importFrom data.table as.data.table copy
+#' @examples
+#' obs <- data.frame(x = 1:3, y = 1:3)
+#' enw_assign_group(obs)
+#' enw_assign_group(obs, by = "x")
 enw_assign_group <- function(obs, by = c()) {
-  if (".group" %in% names(obs)) {
-    stop("Dataset cannot have a column called '.group'.")
+  if (!is.null(obs[[".group"]])) {
+    stop(
+      "The `.group` column is reserved for internal use. Please remove it ",
+      "from your data before calling `enw_assign_group`."
+    )
   }
-  obs <- data.table::as.data.table(obs)
-  if (length(by) == 0) {
-    obs <- obs[, .group := 1]
-  } else {
+  if (length(by) != 0) {
```

Comment:
hmm - feels a bit like the `by` argument should be in add_group. I'm okay with this function overall as a wrapper to the capability (e.g. to manage argument checking that isn't typically done on internal function), but a bit odd this doesn't live in `add_group`.

Reply:
as in `enw_assign_group` and `add_group` could be merged? I agree  (though the does it allow a `.group` var to exist is currently different and needs to be as it is currently used). I think we should discuss in another issue vs this PR though (as here trying to focus on docs).

Reply:
sort of, but not quite - agree another issue. basically, focus the pure technical capabilities in add_group (such as executing on a by clause), but have it assume (because it's internal function) that it's not handed garbage. the enw_assign_group handles all the crap users might throw at it.

Reply:
but this assumes you want to leverage those technical capabilities in other places you're using add_group, not just enw_assign_group

Reply:
ah got you. I think we always want to go via assign group (at least for now) but agree we could have assign group be the exposed to the user version and add_group be the less safe internal use version (though for people inspecting the code this may lead them astray (if for example they want to role their own version of the preprocessing function for some reason).

Reply:
Could you draft up a new issue for this and will discuss/prototype ahead of the next release? I think I need to think about it a little.

Reply:
Added a new issue (#214) to discuss this.